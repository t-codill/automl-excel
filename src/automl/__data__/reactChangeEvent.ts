// tslint:disable: no-any
export const reactChangeEvent: React.ChangeEvent<any> = {
    target: {}
} as unknown as React.ChangeEvent<any>;
