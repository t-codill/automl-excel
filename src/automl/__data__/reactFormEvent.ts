// tslint:disable: no-any
export const reactFormEvent: React.FormEvent<any> = {
    preventDefault(): void { return; }
} as unknown as React.FormEvent<any>;
