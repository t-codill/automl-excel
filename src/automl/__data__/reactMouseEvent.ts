// tslint:disable-next-line: no-any
export const reactMouseEvent = {} as unknown as React.MouseEvent<any>;
