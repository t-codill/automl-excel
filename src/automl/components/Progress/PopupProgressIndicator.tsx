import { Dialog, DialogType, ProgressIndicator } from "office-ui-fabric-react";
import * as React from "react";

interface IPopupProgressIndicatorProps {
    title?: string;
    description?: string;
    percentComplete?: number;
    onDismiss?(): void;
}

export class PopupProgressIndicator
    extends React.PureComponent<IPopupProgressIndicatorProps> {
    public render(): React.ReactNode {

        return <Dialog
            hidden={false}
            minWidth={800}
            onDismiss={this.props.onDismiss}
            modalProps={{
                isBlocking: !this.props.onDismiss
            }}
            dialogContentProps={{
                type: DialogType.normal,
                title: this.props.title
            }}
        >
            <ProgressIndicator
                description={this.props.description}
                percentComplete={this.props.percentComplete} />
        </Dialog>;
    }
}
