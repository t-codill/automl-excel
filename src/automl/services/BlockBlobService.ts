import { TransferProgressEvent } from "@azure/ms-rest-js";
import { BlockBlobURL, Models } from "@azure/storage-blob";

import { blob2string } from "../common/utils/blob2string";
import { csv, ICsvData } from "../common/utils/csv";
import { stream2string } from "../common/utils/stream2string";
import { IServiceBaseProps } from "./ServiceBase";
import { IServiceBaseBlobProps, ServiceBaseBlob } from "./ServiceBaseBlob";

export interface IBlockBlobServiceProps extends IServiceBaseBlobProps {
    blob: Models.BlobItem;
}

const blockSize = 5000;

export class BlockBlobService extends ServiceBaseBlob<BlockBlobURL> {
    private readonly blob: Models.BlobItem;

    constructor(props: IServiceBaseProps, blobProps: IBlockBlobServiceProps) {
        super(props, blobProps, BlockBlobURL, blobProps.blob);
        this.blob = blobProps.blob;
    }

    public async previewBlob(hasHeader: boolean, previewLines: number): Promise<ICsvData | undefined> {
        let content = "";
        const totalLines = previewLines + 1;
        let csvData = csv("", hasHeader, totalLines);
        for (let i = 0; i * blockSize < (this.blob.properties.contentLength || 0); i++) {
            const section = await this.downloadBlob(i * blockSize, blockSize);
            if (section === undefined) {
                return undefined;
            }
            content += section;
            csvData = csv(content, hasHeader, totalLines);
            if (csvData.data.length >= totalLines) {
                break;
            }
        }
        if (csvData.data && csvData.data.length > previewLines) {
            csvData.data.pop();
        }
        return csvData;
    }

    public async uploadBlob(
        file: File,
        percentUpdate: (percent: number) => void
    ): Promise<Models.BlockBlobUploadHeaders | undefined> {
        return this.send(async (client, abortSignal) => {
            return client.upload(this.getAborter(abortSignal), file, file.size, {
                progress: (e: TransferProgressEvent) => { percentUpdate(e.loadedBytes / file.size); }
            });
        });
    }

    public async downloadBlob(from: number, count: number): Promise<string | undefined> {
        const downloadLength = Math.min((this.blob.properties.contentLength || 0) - from, count);
        const content = await this.send(async (client, abortSignal) => {
            return client.download(this.getAborter(abortSignal), from, downloadLength);
        });
        if (!content) {
            return undefined;
        }
        const blob = await content.blobBody;
        if (blob) {
            return blob2string(blob);
        }
        if (content.readableStreamBody) {
            return stream2string(content.readableStreamBody);
        }
        return undefined;
    }
}
