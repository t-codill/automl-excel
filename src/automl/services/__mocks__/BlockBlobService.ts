import { Models } from "@azure/storage-blob";
import { ICsvData } from "../../common/utils/csv";

export class BlockBlobService {
    public async previewBlob(): Promise<ICsvData | undefined> {
        return {
            hasHeader: true,
            header: ["a", "b", "c"],
            data: [{
                a: 1,
                b: 2,
                c: 3
            }],
            delimiter: ","
        };
    }

    public async uploadBlob(): Promise<Models.BlockBlobUploadHeaders | undefined> {
        return {
            requestId: "requestId"
        };
    }

    public async downloadBlob(): Promise<string | undefined> {
        return "Test Blob Content";
    }

    public dispose(): void {
        return;
    }

    public reset(): void {
        return;
    }
}
