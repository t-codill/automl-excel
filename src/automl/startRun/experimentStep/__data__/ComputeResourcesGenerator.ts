import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";

export type ComputeType = "Compute" | "AKS" | "AmlCompute" | "VirtualMachine" | "HDInsight" | "DataFactory" | "Databricks" | "DataLakeAnalytics";

export const getComputeResources = (
    testComputeType: ComputeType,
    testProvisioningState: AzureMachineLearningWorkspacesModels.ProvisioningState | undefined) => {
    let compute: AzureMachineLearningWorkspacesModels.ComputeResource[];
    if (testComputeType === "AmlCompute") {
        compute = [
            {
                name: "Compute1",
                id: "ComputeId1",
                properties: {
                    computeType: testComputeType,
                    provisioningState: testProvisioningState,
                    properties: {
                        scaleSettings: {
                            maxNodeCount: 2,
                            minNodeCount: 0
                        }
                    }
                }
            },
            {
                name: "Compute2",
                id: "ComputeId2",
                properties: {
                    computeType: testComputeType,
                    provisioningState: testProvisioningState,
                    properties: {
                        scaleSettings: {
                            maxNodeCount: 2,
                            minNodeCount: 1
                        }
                    }
                }
            },
            {
                name: "Compute No ID",
                properties: {
                    computeType: testComputeType,
                    provisioningState: testProvisioningState,
                    properties: {
                        scaleSettings: {
                            maxNodeCount: 2,
                            minNodeCount: 1
                        }
                    }
                }
            },
            {
                name: "Compute unusable",
                properties: {
                    computeType: testComputeType,
                    provisioningState: testProvisioningState,
                    properties: {
                        scaleSettings: {
                            maxNodeCount: 2,
                            minNodeCount: 1
                        },
                        nodeStateCounts: {
                            unusableNodeCount: 2
                        }
                    }
                }
            },
            {
                name: "Compute No Property",
                id: "ComputeId2"
            }
        ];
    }
    else {
        compute = [
            {
                name: "Compute1",
                id: "ComputeId1",
                properties: {
                    // tslint:disable-next-line: no-any
                    computeType: testComputeType as any,
                    provisioningState: testProvisioningState
                }
            }
        ];
    }
    return compute;
};
