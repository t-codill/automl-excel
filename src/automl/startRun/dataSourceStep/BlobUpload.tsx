import * as React from "react";
import { BaseComponent } from "../../components/Base/BaseComponent";
import { PopupProgressIndicator } from "../../components/Progress/PopupProgressIndicator";
import { BlockBlobService, IBlockBlobServiceProps } from "../../services/BlockBlobService";

export interface IBlobUploadProps extends IBlockBlobServiceProps {
    file: File;
    cancel(): void;
    complete(): void;
}
export class BlobUpload
    extends BaseComponent<IBlobUploadProps, { percentComplete: number }, {
        blockBlobService: BlockBlobService;
    }> {
    protected readonly serviceConstructors = {
        blockBlobService: BlockBlobService,
    };

    constructor(props: IBlobUploadProps) {
        super(props);
        this.state = {
            percentComplete: 0
        };
    }

    public render(): React.ReactNode {
        return <PopupProgressIndicator
            title="Uploading dataset"
            description={`Uploading ${this.props.file.name} to azure storage.`}
            percentComplete={this.state.percentComplete}
            onDismiss={this.props.cancel}
        />;
    }
    protected readonly getData = async () => {
        const result = await this.services.blockBlobService.uploadBlob(this.props.file, this.updatePercent);
        if (!result) {
            return;
        }
        this.props.complete();
    }

    private readonly updatePercent = (percent: number) => {
        this.setState({ percentComplete: percent });
    }

}
