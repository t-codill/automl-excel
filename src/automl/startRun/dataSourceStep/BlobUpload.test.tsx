import { shallow } from "enzyme";
import * as React from "react";
import { sampleStorageAccount } from "../../../__data__/sampleStorageAccount";
import { sampleStorageBlob } from "../../../__data__/sampleStorageBlob";
import { sampleStorageContainer } from "../../../__data__/sampleStorageContainer";
import { PopupProgressIndicator } from "../../components/Progress/PopupProgressIndicator";
import { BlockBlobService } from "../../services/BlockBlobService";
import { BlobUpload, IBlobUploadProps } from "./BlobUpload";

jest.mock("../../services/BlockBlobService");
let uploadBlobSpy: jest.SpyInstance<ReturnType<BlockBlobService["uploadBlob"]>>;

describe("BlobUpload", () => {
    let cancel: jest.Mock;
    let complete: jest.Mock;
    let props: IBlobUploadProps;
    const sampleFile = new File([], "sample.csv");
    beforeEach(() => {
        uploadBlobSpy = jest.spyOn(BlockBlobService.prototype, "uploadBlob");
        cancel = jest.fn();
        complete = jest.fn();
        props = {
            file: sampleFile,
            account: sampleStorageAccount,
            container: sampleStorageContainer,
            blob: sampleStorageBlob,
            sasToken: "sampleSasToken",
            cancel,
            complete
        };
    });

    it("should render", async () => {
        const tree = shallow(<BlobUpload {...props} />);
        await Promise.resolve();
        expect(uploadBlobSpy)
            .toBeCalledWith(sampleFile, expect.anything());
        expect(tree)
            .toMatchSnapshot();
        expect(complete)
            .toBeCalled();
    });

    it("should not render if upload canceled", async () => {
        uploadBlobSpy.mockReturnValue(Promise.resolve(undefined));
        shallow(<BlobUpload {...props} />);
        await Promise.resolve();
        expect(complete)
            .not
            .toBeCalled();
    });

    it("should update Percent", async () => {
        uploadBlobSpy.mockImplementation(async (_: File, updatePercent: (percent: number) => void) => {
            updatePercent(12);
            return Promise.resolve(undefined);
        });
        const tree = shallow(<BlobUpload {...props} />);
        await Promise.resolve();
        expect(tree.find(PopupProgressIndicator)
            .prop("percentComplete"))
            .toBe(12);
    });
});
