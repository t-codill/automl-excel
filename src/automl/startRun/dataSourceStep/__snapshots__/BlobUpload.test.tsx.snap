// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlobUpload should render 1`] = `
<PopupProgressIndicator
  description="Uploading sample.csv to azure storage."
  onDismiss={[MockFunction]}
  percentComplete={0}
  title="Uploading dataset"
/>
`;
